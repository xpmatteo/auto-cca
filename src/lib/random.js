
export function randomShuffleArray(array) {
    for (let i = 0; i < array.length; i++) {
        const j = Math.floor(Math.random() * (array.length - i)) + i;
        [array[i], array[j]] = [array[j], array[i]];
    }
}

/**
 * @template T
 * @param {T[]} array
 * @returns {T}
 */
export function randomElement(array) {
    return array[Math.floor(Math.random() * array.length)];
}

const RANDOM_VALUES = [0.1, 0.3, 0.2, 0.7, 0.9, 0.6, 0.4, 0.8];
let nextRandom = 0;
export function fixedRandom() {
    return RANDOM_VALUES[nextRandom++ % 8];
}
