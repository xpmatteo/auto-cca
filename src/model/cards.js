import { OrderUnitsPhase } from "./phases/order_units_phase.js";
import { RESULT_HEAVY, RESULT_LIGHT, RESULT_MEDIUM } from "./dice.js";
import { Point } from "../lib/hexlib.js";
import { MovementPhase } from "./phases/MovementPhase.js";
import { BattlePhase } from "./phases/BattlePhase.js";
import { FirePhase } from "./phases/fire_phase.js";

/*
    Cards are used to order units during the order phase.
    They determine the structure of the turn.
    Most turns will have a MovementPhase followed by a BattlePhase, but some cards
    have different turn structures.

    Cards are stateless and immutable.
*/

export const CARD_IMAGE_SIZE = new Point(400, 560);
class Card {
    phases(game) {
        return [
            new MovementPhase(),
            new BattlePhase(),
        ];
    }
}

export class OrderHeavyTroopsCard extends Card {
    name = "Order Heavy Troops";
    url = "images/cards/Order Heavy Troops.gif";
    orderPhase(game) { return new OrderUnitsPhase(game.commandSize(), RESULT_HEAVY); }
}

export class OrderMediumTroopsCard extends Card {
    name = "Order Medium Troops";
    url = "images/cards/Order Medium Troops.gif";
    orderPhase(game) { return new OrderUnitsPhase(game.commandSize(), RESULT_MEDIUM); }
}

export class OrderLightTroopsCard extends Card {
    name = "Order Light Troops";
    url = "images/cards/Order Light Troops.gif";
    orderPhase(game) { return new OrderUnitsPhase(game.commandSize(), RESULT_LIGHT); }
}

export class MoveFireMoveCard extends Card {
    name = "Move Fire Move";
    url = "images/cards/Move-Fire-Move.gif";
    orderPhase(game) { return new OrderUnitsPhase(game.commandSize(), RESULT_LIGHT); }
    phases(game) {
        return [
            new MovementPhase("1st movement"),
            new FirePhase(),
            new MovementPhase("2nd movement"),
        ];
    }
}
