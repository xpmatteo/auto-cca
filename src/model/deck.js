
import { randomShuffleArray } from "../lib/random.js";
import {
    Card, MOVE_FIRE_MOVE_CARD,
    ORDER_2_CENTER_CARD,
    ORDER_2_LEFT_CARD,
    ORDER_2_RIGHT_CARD,
    ORDER_3_CENTER_CARD,
    ORDER_3_LEFT_CARD,
    ORDER_3_RIGHT_CARD,
    ORDER_4_CENTER_CARD,
    ORDER_4_LEFT_CARD,
    ORDER_4_RIGHT_CARD, ORDER_HEAVY_TROOPS_CARD,
    ORDER_LIGHT_TROOPS_CARD,
    ORDER_MEDIUM_TROOPS_CARD
} from './cards.js';

function repeat(number, item) {
    return Array(number).fill(item);
}

export function deckSpecToCards(deckSpec) {
    const result = [];
    for (const [number, card] of deckSpec) {
        result.push(...repeat(number, card));
    }
    return result;
}

export class Deck {
    /** @type {Card[]} */
    #talon;
    /** @type {Card[]} */
    #discards = [];

    /**
     * @param {Card[]} cards
     * @param {Card[]} discards
     */
    constructor(cards, discards = []) {
        this.#talon = cards;
        this.#discards = discards;
    }

    /** @returns {Card[]} */
    talon() {
        return this.#talon.slice();
    }

    /** @returns {Card[]} */
    discards() {
        return this.#discards.slice();
    }

    /** @returns {number} */
    get talonSize() {
        return this.#talon.length;
    }

    /** @returns {number} */
    get discardsSize() {
        return this.#discards.length;
    }

    /** @returns {void} */
    shuffle() {
        this.#talon = randomShuffleArray(this.#talon.concat(this.#discards));
        this.#discards = [];
    }

    /**
     * @param {number} number
     * @returns {Card[]}
     */
    deal(number) {
        const cards = this.#talon.slice(0, number);
        this.#talon = this.#talon.slice(number);
        return cards;
    }

    /**
     * @param {Card} card
     * @returns {void}
     */
    discard(card) {
        this.#discards.push(card);
    }

    clone() {
        return new Deck(this.#talon.slice(), this.#discards.slice());
    }
}

/** @type {[number,Card][]} */
const DECK_SPEC = [
    [3, ORDER_2_LEFT_CARD],
    [4, ORDER_2_CENTER_CARD],
    [3, ORDER_2_RIGHT_CARD],
    [3, ORDER_3_LEFT_CARD],
    [4, ORDER_3_CENTER_CARD],
    [3, ORDER_3_RIGHT_CARD],
    [1, ORDER_4_LEFT_CARD],
    [1, ORDER_4_CENTER_CARD],
    [1, ORDER_4_RIGHT_CARD],
    [4, ORDER_LIGHT_TROOPS_CARD],
    [3, ORDER_MEDIUM_TROOPS_CARD],
    [2, ORDER_HEAVY_TROOPS_CARD],
    [2, MOVE_FIRE_MOVE_CARD],
];

export const THE_DECK = new Deck(deckSpecToCards(DECK_SPEC));

