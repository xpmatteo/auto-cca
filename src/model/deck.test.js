import { fixedRandom, resetFixedRandom } from "../lib/random.js";
import { ORDER_2_CENTER_CARD, ORDER_4_RIGHT_CARD, ORDER_HEAVY_TROOPS_CARD } from "./cards.js";
import { Deck, deckSpecToCards, THE_DECK } from "./deck.js";

describe('The deck!', () => {
    const originalRandom = Math.random;
    beforeEach(() => {
        resetFixedRandom();
        Math.random = fixedRandom;
    });

    afterEach(() => {
        Math.random = originalRandom;
    });

    const deckSpec = [[2, ORDER_4_RIGHT_CARD], [3, ORDER_HEAVY_TROOPS_CARD]];

    test('deck construction', () => {
        const deck = new Deck(deckSpecToCards(deckSpec));

        expect(deck.talon()).toEqual([
            ORDER_4_RIGHT_CARD,
            ORDER_4_RIGHT_CARD,
            ORDER_HEAVY_TROOPS_CARD,
            ORDER_HEAVY_TROOPS_CARD,
            ORDER_HEAVY_TROOPS_CARD,
        ]);
    });

    test('full deck cards', () => {
        expect(THE_DECK.talonSize).toBe(34);
        expect(THE_DECK.discardsSize).toBe(0);
        expect(THE_DECK.talon().toString()).toMatchSnapshot();
    });

    describe('shuffling', () => {
        test('no discards', () => {
            const deck = new Deck(deckSpecToCards(deckSpec));

            deck.shuffle();

            expect(deck.talon().map(c => c.name)).toEqual([
                "Order Four Units Right",
                "Order Heavy Troops",
                "Order Four Units Right",
                "Order Heavy Troops",
                "Order Heavy Troops",
            ]);
        });

        test('with discards', () => {
            const deck = new Deck(deckSpecToCards(deckSpec));
            deck.discard(ORDER_2_CENTER_CARD);

            deck.shuffle();

            expect(deck.discardsSize).toBe(0);
            expect(deck.talon().map(c => c.name)).toEqual([
                "Order Four Units Right",
                "Order Heavy Troops",
                "Order Four Units Right",
                "Order Two Units Center",
                "Order Heavy Troops",
                "Order Heavy Troops",
            ]);
        });
    });

    test('dealing cards when talon is sufficient', () => {
        const deck = new Deck(deckSpecToCards(deckSpec));

        const [card1, card2, card3] = deck.deal(3);

        expect(deck.talonSize).toBe(2);
        expect(deck.discardsSize).toBe(0);
        expect(card1.name).toBe("Order Four Units Right");
        expect(card2.name).toBe("Order Four Units Right");
        expect(card3.name).toBe("Order Heavy Troops");
        expect(deck.talon().map(c => c.name)).toEqual([
            "Order Heavy Troops",
            "Order Heavy Troops",
        ]);
    });

    test('discarding', () => {
        const deck = new Deck(deckSpecToCards(deckSpec));

        deck.discard(ORDER_4_RIGHT_CARD);

        expect(deck.talonSize).toBe(5);
        expect(deck.discardsSize).toBe(1);
    });

    test('clone', () => {
        const deck = new Deck(deckSpecToCards(deckSpec));
        deck.discard(ORDER_2_CENTER_CARD);

        const clone = deck.clone();

        expect(clone.talon()).toEqual(deck.talon());
        expect(clone.discards()).toEqual([ORDER_2_CENTER_CARD]);
        expect(clone).not.toBe(deck);
    });
});
