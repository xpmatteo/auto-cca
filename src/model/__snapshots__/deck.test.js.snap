// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The deck! full deck cards 1`] = `"Order Two Units Left,Order Two Units Left,Order Two Units Left,Order Two Units Center,Order Two Units Center,Order Two Units Center,Order Two Units Center,Order Two Units Right,Order Two Units Right,Order Two Units Right,Order Three Units Left,Order Three Units Left,Order Three Units Left,Order Three Units Center,Order Three Units Center,Order Three Units Center,Order Three Units Center,Order Three Units Right,Order Three Units Right,Order Three Units Right,Order Four Units Left,Order Four Units Center,Order Four Units Right,Order Light Troops,Order Light Troops,Order Light Troops,Order Light Troops,Order Medium Troops,Order Medium Troops,Order Medium Troops,Order Heavy Troops,Order Heavy Troops,Darken the Sky,Move Fire Move,Move Fire Move"`;
